compiler/nilcheck.nim

```diff
@@ -0,0 +1,928 @@
+    for child in n:
+      if directStop(child):
+        return true
+  of nkReturnStmt, nkBreakStmt, nkContinueStmt, nkRaiseStmt:
+    return true
+  of nkIfStmt, nkElse:
+    return false
+  else:
+    aecho n.kind
+  return false
+
+proc reverse(value: Nilability): Nilability =
+  case value:
+  of Nil: Safe
+  of MaybeNil: MaybeNil
+  of Safe: Nil
+
+proc reverse(kind: TransitionKind): TransitionKind =
+  case kind:
+  of TNil: TSafe
+  of TSafe: TNil
+  else: raise newException(ValueError, "expected TNil or TSafe")
+
+proc reverseDirect(map: NilMap): NilMap =
+  result = map.copyMap()
+  for graphIndex, local in result.locals:
+    result.locals[graphIndex] = reverse(local)
+    result.history[graphIndex][^1].kind = reverse(result.history[graphIndex][^1].kind)
+    result.history[graphIndex][^1].nilability = result.locals[graphIndex]
+
+proc checkCondition(n, ctx, map; reverse: bool, base: bool): NilMap =
+  ## check conditions : used for if, some infix operators
+  ##   isNil(a)
+  ##   it returns a new map: you need to reverse all the direct elements for else
+
+  # check that the condition is valid
+  
+  #if base:
+  #  map.base = map
+  # TODO dont inc abstractTime for `of` / `else` artificcial nodes
+  if n.kind == nkCall:
+    inc ctx.abstractTime # nkCall
+    # echo "nilcheck : abstractTime " & $ctx.abstractTime & " nkCall " & $n
+
+    result = newNilMap(map, map.base)
+    for element in n:
+      result = check(element, ctx, result).map
+
+    if n[0].kind == nkSym and n[0].sym.magic == mIsNil:    
+      # result = newNilMap(map, map.base) # if base: map else: map.base)
+
+      let a = ctx.graph(n[1])
+      result.store(a, if not reverse: Nil else: Safe, if not reverse: TNil else: TSafe, n.info, n)
+    else:
+      discard
+  elif n.kind == nkPrefix and n[0].kind == nkSym and n[0].sym.magic == mNot:
+    inc ctx.abstractTime # nkPrefix
+    # echo "nilcheck : abstractTime " & $ctx.abstractTime & " nkPrefix " & $n
+    result = checkCondition(n[1], ctx, map, not reverse, false)
+  elif n.kind == nkInfix:
+    inc ctx.abstractTime # nkInfix
+    # echo "nilcheck : abstractTime " & $ctx.abstractTime & " nkInfix " & $n
+    result = newNilMap(map, map.base)
+    result = checkInfix(n, ctx, result).map
+  else:
+    result = check(n, ctx, map).map
+    result = newNilMap(map, map.base)
+  assert not result.isNil
+  assert not result.previous.isNil
+
+proc checkResult(n, ctx, map) =
+  let resultNilability = map[ctx.graph(resultId)]
+  case resultNilability:
+  of Nil:
+    message(ctx.config, n.info, warnStrictNotNil, "return value is nil")
+  of MaybeNil:
+    message(ctx.config, n.info, warnStrictNotNil, "return value might be nil")
+  of Safe:
+    discard    
+
+proc checkBranch(n, ctx, map): Check =
+  result = check(n, ctx, map)
+
+
+# Faith!
+
+proc check(n: PNode, ctx: NilCheckerContext, map: NilMap): Check =
+  if map.isNil:
+    localError(ctx.config, n.info, "map is nil: something went wrong in nilcheck")
+    quit 1
```

Comment:
No ``quit`` here.

Reply:
should I just assert?

Reply:
or just wait and use NilMap not nil later